### PR TITLE
TMA doc: part 3

### PR DESCRIPTION
**Pull request by zasdfgbnm**
- `tma.md`
  - Define box by compositing
  - Step 3: define tile 
- `tma-modeling-in-depth.md`
  - The unachievability of strong correctness for indivisible element stride
  - The lowering strategy

### CODE REVIEW

File: doc/reading/tma-modeling-in-depth.md [hunk 1]
```diff
@@ -168,6 +176,298 @@ Another example is the following Figure 4:
...
+
+$\square$
+
+</details>
+
+Now, let's go back to the example in Figure 5.
+In this example, all the conditions of Theorem 6 are satisfied for the outer dimension,
+so strong correctness is unachievable. What if some condition is violated?
+
+If the element stride divided the box size, then the striding split would not create any hole at all.
```
Comment (naoyam) [inline doc/reading/tma-modeling-in-depth.md:424]: What's the "striding split'?

Comment (zasdfgbnm) [reply doc/reading/tma-modeling-in-depth.md:424]: It is in `tma.md` `Step 3: define tile`

```
For strided tile, we do an inner-split on the box IterDomain by the *element stride*.
We call this split "*striding split*", the inner output of this split "*stride IterDomain*",
```

File: doc/dev/tma.md [hunk 2]
```diff
@@ -86,12 +90,87 @@ We call this split expression a "*boxing split*", the input of this split a "*pa
...
 For the case of Figure 1, if both box dimensions are defined by partitioning,
-the schedule should look like the Figure 3 below:
+the schedule should look like the Figure 4 below:
 
-![Figure 3: Boxing by partitioning](tma/box-by-partitioning.svg)
+![Figure 4: Boxing by partitioning](tma/box-by-partitioning.svg)
 
 Please note that, although in the above example, the split is divisible, this does not have to be the case in general.
 
 #### Define box by compositing
```
Comment (zasdfgbnm) [inline doc/dev/tma.md:99]: This is supported in https://github.com/NVIDIA/Fuser/pull/1991.

File: doc/dev/tma.md [hunk 3]
```diff
@@ -86,12 +90,87 @@ We call this split expression a "*boxing split*", the input of this split a "*pa
...
+The TMA domain of the tensor has nine IterDomains.
+The contiguities of these nine IterDomains are `(T, T, T, F, T, T, T, T, T)`.
+We scheduled a 5D TMA operation, where in these five dimensions, one of them defines box by partitioning, and others define box by compositing.
+Green color means box IterDomain and brown color means coordinate IterDomain.
+The first dimension is `[I1, I2]`, where `I1` is coordinate IterDomain and `I2` is box IterDomain.
+The second dimension is `I3`, it is the only dimension that defines box by partitioning.
+`I3` is the partitioned IterDomain, `I11` is the box IterDomain, and `I10` is the coordinate IterDomain.
+The third dimension is `I4`, it has no box IterDomain, therefore, the box size is implicitly one.
+The fourth dimension is `[I5, I6]`, it has no coordinate IterDomain, therefore this dimension only has one box,
+whose size is the produce of the extents of `I5` and `I6`.
```
Comment (naoyam) [inline doc/dev/tma.md:143]:
```suggestion
whose size is the product of the extents of `I5` and `I6`.
```

File: doc/dev/tma.md [hunk 4]
```diff
@@ -86,12 +90,87 @@ We call this split expression a "*boxing split*", the input of this split a "*pa
...
+The fifth dimension is `[I7, I8, I9]`, where `I8` and `I9` are box IterDomain, and `I7` is coordinate IterDomain.
+The imaginary TMA domain has five IterDomains, they are `merge(I1, I2)`, `I3`, `I4`, `merge(I5, I6)`, and `merge(I7, I8, I9)`.
+
+### Step 3: define tile
+
+TMA's tile can be dense or strided.
+Most commonly, we use dense tile.
+For dense tile, we define the *tile IterDomain* as the box IterDomain,
+and use the word *tile* and *box* interchangably.
```
Comment (naoyam) [inline doc/dev/tma.md:152]:
```suggestion
and use the word *tile* and *box* interchangeably.
```

File: doc/dev/tma.md [hunk 5]
```diff
@@ -86,12 +90,87 @@ We call this split expression a "*boxing split*", the input of this split a "*pa
...
 
-![Figure 3: Boxing by partitioning](tma/box-by-partitioning.svg)
+![Figure 4: Boxing by partitioning](tma/box-by-partitioning.svg)
 
 Please note that, although in the above example, the split is divisible, this does not have to be the case in general.
 
 #### Define box by compositing
 
-TODO: this documentation is under construction
+Define box by compositing is to select a slice of IterDomains from the TMA domain and consider the innermost IterDomain(s) in the slice as box.
```
Comment (naoyam) [inline doc/dev/tma.md:101]: If the `I5` TMA domain is scheduled, then can we replace this way of defining boxes by the above partitioning method? I wonder why we need this method in addition to the partitioning.

Comment (zasdfgbnm) [reply doc/dev/tma.md:101]: Yes, we can always replace this way of defining boxes by partitioning. This way exists primarily for convenience. For example, if I have a discontiguous tensor [I1, I2], then you will be forced to partition both I1 and I2. But what often happen is, you only want to partition I2, and use box size 1 for I1. With this way of defining box, we will be able to automatically consider this as a size 1 box. Another example is, when you want to write a unit test, you often want to treat the entire dim as box, with only partitioning, you are forced to split this dim by its extent to get a [1, N]. But with this mechanism, you will be able to just parallelize this dim as `ParallelType::Bulk`.

File: doc/reading/tma-modeling-in-depth.md [hunk 6]
```diff
@@ -168,6 +176,298 @@ Another example is the following Figure 4:
...
+If the the tensor shape was `(4, 2)` instead of `(8, 2)`,
+which violated the condition that the box size is smaller than the tensor size,
+then the red items in Figure 6 would not exist, because items from `8` to `15` would now be out of boundary.
+
+If the element stride was `5` instead of `3`,
+the red items in Figure 6 would still exist without extra protextion,
+because the example of $i_3 = 0$, $i_4 = 4$, and $i_1 = 4$ mentioned above is still valid.
+However, by lowering smartly, we can add a check for the range of `I6` (i.e. $i_6 < 4$) to get rid of the red items.
+
+## The lowering strategy
```
Comment (naoyam) [inline doc/reading/tma-modeling-in-depth.md:435]: This part is still difficult for me to follow. Some examples would be helpful.

File: doc/reading/tma-modeling-in-depth.md [hunk 7]
```diff
@@ -168,6 +176,298 @@ Another example is the following Figure 4:
...
+
+Theorem 6 is so powerful that it deserve a fancier name "**The fundamental theorem of TMA correctness**",
+or **FTTC** in short.
+It is powerful because the condition in it is both necessary and sufficient,
+and it directly tells us the correct lowering strategy:
+
+We should check the condition of FTTC.
+If the condition is true, then it is impossible to achieve strong correctness.
+So raise an error if strong correctness is required.
+If strong correctness is not required, generate code like below:
```
Comment (naoyam) [inline doc/reading/tma-modeling-in-depth.md:445]: If the strong correctness is not required, it means the weak correctness is sufficient, right? If so, why does it need additional predicates? Doesn't TMA automatically guarantee the weak correctness?

File: doc/reading/tma-modeling-in-depth.md [hunk 8]
```diff
@@ -168,6 +176,298 @@ Another example is the following Figure 4:
...
+
+```C++
+if (predicates_for_all_non_TMA_protected_IterDomains_except_box_IterDomains) {
+  // Do TMA normally
+  tma(...);
+}
+```
+
+If the condition of FTTC is false, then we always achieve strong correctness by generate the following code:
```
Comment (naoyam) [inline doc/reading/tma-modeling-in-depth.md:454]: Also here, FTTC is false, which means the strong correctness is guaranteed. Why are these predicates necessary?

File: doc/reading/tma-modeling-in-depth.md [hunk 9]
```diff
@@ -168,6 +176,298 @@ Another example is the following Figure 4:
...
+
+![Figure 6: Indivisible strided tile data](tma-modeling-in-depth/no-strong-correctness-data.svg)
+
+From the above figure, we see that there are two tiles where the first half tile is blue and the second half tile is red.
+However, the granularity on TMA that programmer has control of is per tile.
+There is no way for a programmer to tell the hardware that they want part of a tile to be filled with data but part of it filled with zero even if they are not out of the boundary of global memory.
+
+The above observation leads to the following theorem:
+
+**Theorem 6:** It is impossible to achieve strong correctness if and only if there exist a dimension that satisfies all of the following conditions:
```
Comment (naoyam) [inline doc/reading/tma-modeling-in-depth.md:203]: nit: perhaps easier to remember to rephrase this as:
 "The strong correctness is achieved only if the following conditions are satisfied."

File: doc/reading/tma-modeling-in-depth.md (hunk 7 above)
Comment (zasdfgbnm) [reply doc/reading/tma-modeling-in-depth.md:445]: It does automatically guarantee the weak correctness, but if there is a way to save some memory traffic, why not do it?

Comment (naoyam) [reply doc/reading/tma-modeling-in-depth.md:445]: I thought the only concern here was correctness.

File: doc/reading/tma-modeling-in-depth.md (hunk 9 above)
Comment (zasdfgbnm) [reply doc/reading/tma-modeling-in-depth.md:203]: This is not what Theorem 6 is saying. What Theorem 6 says is, if the condition is satisfied, then no matter what you do, you can not achieve strong correctness. If the condition is not satisfied, then you can achieve strong correctness by doing something.

So there are three levels:
- If there is no indivisible splits on TMA-unprotected IDs, then strong correctness is automatically achieved without needing to do anything
- If there are indivisible splits on TMA-unprotected IDs, but the condition of FTTC is not satisfied, then strong correctness is achievable by predicating these TMA-unprotected IDs.
- If the condition of FTTC is satisfied, then you have no choice but to raise an error.

Comment (naoyam) [reply doc/reading/tma-modeling-in-depth.md:203]: What I wanted to suggest was also to change the conditions because it sounded like a double negative. For example, the following levels would be phrased like:

> but the condition of FTTC is satisfied, then strong correctness is achievable by predicating these TMA-unprotected IDs.

> If the condition of FTTC is not satisfied, then you have no choice but to raise an error.

Just a nit suggestion

File: doc/reading/tma-modeling-in-depth.md (hunk 7 above)
Comment (zasdfgbnm) [reply doc/reading/tma-modeling-in-depth.md]: Modified the doc to be more clear on that

File: doc/reading/tma-modeling-in-depth.md (hunk 8 above)
Comment (zasdfgbnm) [reply doc/reading/tma-modeling-in-depth.md]: Added a table clarifying

File: doc/reading/tma-modeling-in-depth.md (hunk 9 above)
Comment (zasdfgbnm) [reply doc/reading/tma-modeling-in-depth.md:203]: I am not sure, but if rephrased like "Strong correctness is achievable if no dim satisfies both of the following conditions: 1, 2", then when we say "the condition of theorem 6", it is hard to tell if it refers to "no dim satisfies 1 and 2" or refers to "1, 2".

**Review by naoyam — APPROVED** [review]
Thanks for the clarifications. Much more clear to me.
